google/cloud/bigtable/completion_queue.h

```diff
@@ -76,6 +77,35 @@ class CompletionQueue {
     return MakeDeadlineTimer(std::chrono::system_clock::now() + duration);
   }
 
+  /**
+   * Make an asynchronous unary RPC.
+   *
+   * @param async_call a callable to start the asynchronous RPC.
+   * @param request the contents of the request.
+   * @param context an initialized request context to make the call.
+   *
+   * @tparam AsyncCallType the type of @a async_call. It must meet be invocable
+   *     with `(grpc::ClientContext*,
+   *     requirements for `internal::CheckAsyncUnaryRpcSignature<>`.
+   * @tparam Request the type of the request parameter in the gRPC.
+   *
+   * @return a future that becomes satisfied when the operation completes.
+   */
+  template <
+      typename AsyncCallType, typename Request,
+      typename Sig = internal::AsyncCallResponseType<AsyncCallType, Request>,
+      typename Response = typename Sig::type,
+      typename std::enable_if<Sig::value, int>::type = 0>
```

Comment:
I was wondering what this enabler was doing. If `Sig` is an `AsyncCallResponseType`, it was not clear to me what `Sig::value` meant. It looks like the value member is inherited from `false_type` or `true_type`. So the `AsyncCallResponseType` struct is both a meta-"function" (returning the response type as `::type`) and also a meta-"predicate" setting `::value` to true or false depending on something.

I don't know if I've seen this style of metafunction that is also a predicate; I usually see one or the other. Are there advantages to doing this over relying on sfinae? For example, this template signature could be written as:

```cc
template <
    typename AsyncCallType, typename Request,
    typename Response = typename internal::AsyncCallResponseType<AsyncCallType, Request>::type>
```

if the metafunctions were written such that the primary template did not define `::type` when the signature is incorrect and therefore there is no type. Doing this kind of makes sense, since there really is no async response type when the wrong signature is used. For example:

```cc
template <typename ResponseType>
struct AsyncCallResponseTypeUnwrap {};  // No ::type member alias

template <typename ResponseType>
struct AsyncCallResponseTypeUnwrap<
    std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<ResponseType>>> {
  using type = ResponseType;
};
```

No change is required, and I'm not arguing *for* this. Just curious to hear your thoughts about the pros-n-cons of each.

Reply:
> Just curious to hear your thoughts about the pros-n-cons of each.

Just using SFINAE is simpler to write, of course.  With the predicate and the `std::enable_if<>` Clang produces nice error messages (or would, if I was good with names and AsyncCallResponseTypeUnwrap had a decent name):

```
  auto future = cq.MakeUnaryRpc(
                ~~~^~~~~~~~~~~~
/v/google/cloud/bigtable/completion_queue.h:104:30: note: candidate template ignored: requirement 'AsyncCallResponseTypeUnwrap<void>::value' was not satisfied [with AsyncCallType = (lambda at /v/google/cloud/bigtable/completion_queue_test.cc:218:7), Request = google::bigtable::admin::v2::GetTableRequest, Sig = google::cloud::bigtable::v0::internal::AsyncCallResponseTypeUnwrap<void>, Response = void]
  future<StatusOr<Response>> MakeUnaryRpc(
```

Reply:
Gotcha. Thanks. There are already a bunch of other meta functions following this pattern, so I do not think you should change anything.

But for the sake of discussion and my own clarification, I'll reply w/ a followup comment/question.

I'm totally in favor of using std::enable_if to produce better error messages. I've seen this have the most improvement when the use of enable_if moves the sfinae error from the *body* of the template instantiation and into the template *parameter list*. In that case, using std::enable_if greatly improves the compiler error messages.

However, in this case, we're trading one sfinae error within the template parameter list for another sfinae error within the template parameter list. This does change the generated error slightly, but it's not clear to me whether it's an improvement or not.

Experimenting with https://godbolt.org/z/i-uBob, I see an error message difference like:

**Error generated using `std::enable_if<>`**
```
<source>:34:11: error: no matching function for call to 'Call'
  return Call<decltype(Bad)>();
         ^~~~~~~~~~~~~~~~~~~
<source>:28:5: note: candidate template ignored: requirement 'AsyncCallResponseTypeUnwrap<int>::value' was not satisfied [with F = int (), Sig = AsyncCallResponseTypeUnwrap<int>, Response = void]
int Call() { return 2; }
   ^
1 error generated.
Compiler returned: 1
```

**Error generated with vanilla SFINAE in the template parameter list**
```
<source>:34:11: error: no matching function for call to 'Call'
  return Call<decltype(Bad)>();
         ^~~~~~~~~~~~~~~~~~~
<source>:28:5: note: candidate template ignored: substitution failure [with F = int ()]: implicit instantiation of undefined template 'AsyncCallResponseTypeUnwrap<int>'
int Call() { return 2; }
   ^
1 error generated.
Compiler returned: 1
```

The error from the second is shorter, but it's hard to tell if either one is clearer.